mlx/backend/metal/indexing.cpp

```diff
@@ -142,7 +142,45 @@ void Scatter::eval_gpu(const std::vector<array>& inputs, array& out) {
   // Get kernel name
   std::ostringstream kname;
   std::string idx_type_name = nidx ? type_to_name(inputs[1]) : "";
-  kname << "scatter" << type_to_name(out) << idx_type_name;
+
+  int idx_ndim = nidx ? inputs[1].ndim() : 0;
+  bool index_nd1_specialization = (idx_ndim == 1);
+
+  // Bail from fast path (1d index specialization) if scatter dims aren't
+  // the outermost dims and contiguous since update access won't be raster
+  // order.
+  for (auto i = 0; i < axes_.size() && index_nd1_specialization; i++) {
+    if (axes_[i] != i) {
+      index_nd1_specialization = false;
+      break;
+    }
+  }
+
+  auto check_for_broadcast_dims =
+      [](const std::vector<size_t>& strides) -> bool {
+    for (auto i = 0; i < strides.size(); i++) {
+      if (strides[i] == 0) {
+        return false;
+      }
+    }
+    return true;
+  };
+
+  // Bail from fast path (1d index specialization) if any of the dims are
+  // broadcasted, since we can't rely on linear indexing in that case.
+  for (int i = 1; i < inputs.size() && index_nd1_specialization; i++) {
+    if (check_for_broadcast_dims(inputs[i].strides()) == false) {
+      index_nd1_specialization = false;
+      break;
+    }
+  }
```

Comment:
Does it matter if one array is col contiguous and another is row contiguous? This will pass if e.g. one index is col contiguous (F-order) and another is row-contiguous (C-order). But you might not want that?

Reply:
I think maybe what you want is all the arrays are `row_contiguous` (e.g. `inputs[i].flags().row_contiguous == True`).

Reply:
Also I'm wondering if it's really so expensive to just include the stride? it's just one extra multiplication.. and it would make this kernel way more general.